rclpy/test/test_logging_rosout.py

```diff
@@ -60,26 +63,26 @@ def test_enable_rosout(
     )
     executor.add_node(node)
 
-    global raw_subscription_msg
-    raw_subscription_msg = None
+    global receive_expected_rosout_msg
+    receive_expected_rosout_msg = False
     # create subscriber of 'rosout' topic
     node.create_subscription(
         Log,
         '/rosout',
-        raw_subscription_callback,
-        1,
-        raw=True
+        rosout_subscription_callback,
+        1
     )
 
-    node.get_logger().info('SOMETHING')
-    executor.spin_once(timeout_sec=1)
+    cycle_count = 0
+    while cycle_count < 5 and receive_expected_rosout_msg is False:
+        node.get_logger().info('SOMETHING')
+        cycle_count += 1
+        executor.spin_once(timeout_sec=1)
```

Comment:
Looping like this looks a lot more robust to me, thanks.

I am curious though about the structure and the decision to try logging multiple times. I think it's ok for the test to assert that a log message is received (it is supposed to be delivered reliably), though. Why not just keep spinning, rather than resending the logging each time.

One reason I could see for this is that this doesn't wait for the subscription to be "matched" to the rosout publisher, so it might "miss" the initial logging message.

There are two ways you could try to improve this I think. You could either:

- loop until you receive a log on the subscription, resending each loop as you do now
  - this ensures communication has been established
- then send one log message, and loop until the subscription receives that exact message or a time elapses
  - assert that the exact message was received

Or you could do something similar except you use the matched status as a precondition for checking for the message:

- wait for the subscription (and/or the publisher) to be matched
- send on log message, and loop until it is received

----

Separately, I'd recommend looping based purely on a time limit, rather than a cycle count, because you cannot ensure that there are not < 5 events to be handled when you enter this loop. The implementation may queue up as many events as is needed prior to this...

Reply:
** not ensure that there are not < 5 events to be handled **, this is a little strange. Maybe in the future, these will be some build-in events which need to be handled if user add a node into executor.  

Based on understand your comments, implementing the codes likes below

```python
    max_difference_time = 5
    begin_time = time.time()
    while rosout_subscription_msg is None:
        node.get_logger().info('SOMETHING')
        executor.spin_once(timeout_sec=1)
        if int(time.time() - begin_time) > max_difference_time:
            break
```

Reply:
Yeah, that looks right, but I'd combine the while loop and break condition, to something like:

```python
while rosout_subscription_msg is None and time_not_expired:
    # ...
```

Just because it is easier to read.

Reply:
Agree. Update codes.